Some changes to the library

yombunker opened the pull request:
- Trying to clean up some of the code by using @IntDef
- Moving the "sequencing" code to different classes
   - to avoid having to "switch" on it every time we need a new position.
- Adding TODOs to fix existing bug
- marking as final some variables
- refactoring some nested IFs
- fixing bug where not checking if is animated inside the IF query

yombunker committed:
- Removed unnecessary gradle commands
- Added annotations
- Simplified the complexity of some methods
- Added: Todo to fix bug

yombunker committed:
- Added IntDef for the different types of animations
- Added IntDef for the different types of backgrounds
- Added IntDef for the different types of sequence order

yombunker committed:
- Added sequence generators

yombunker committed:
- Small changes to DotViews

uDevel commented on typing-indicator/src/main/java/com/udevel/widgetlab/BouncingSlidingDotView.java:
```diff
@@ -204,8 +199,8 @@ public void onAnimationUpdate(ValueAnimator animator) {
-        for (int i = 1; i <= indexToParent; i++) {
-
+        int i = indexToParent;
+        while (i-- > 0) {
```
how is this better?

yombunker replied:
a little formatting and you can see this "while (i --> 0)" which just tells you that you are going from i to 0, but, if you don't like it i can revert it :)

uDevel commented on typing-indicator/src/main/java/com/udevel/widgetlab/sequences/RandomNoRepetitionSequence.java:
```diff
@@ -0,0 +1,16 @@
+    public int nextIndex(int currentIndex, int numberOfElements) {
+        int nextValue;
+        do {
+            nextValue = super.nextIndex(currentIndex, numberOfElements);
```
why need to call super?

uDevel commented on typing-indicator/src/main/java/com/udevel/widgetlab/sequences/RandomNoRepetitionSequence.java:
```diff
@@ -0,0 +1,16 @@
+    public int nextIndex(int currentIndex, int numberOfElements) {
+        int nextValue;
+        do {
+            nextValue = super.nextIndex(currentIndex, numberOfElements);
```
For fun, this can potentially be infinite loop.  :)

yombunker replied:
i thought about it as well, but, because it uses a normal distribution it "shouldn't" happen, but, i guess you are correct technically, i was thinking of adding a counter and then just finishing if it took more than like 5 tries or something.

uDevel commented on typing-indicator/src/main/java/com/udevel/widgetlab/TypingIndicatorView.java:
```diff
@@ -15,108 +18,90 @@
 
-            }
+    private SequenceGenerator sequenceGenerator;
+    private int currentAnimateDotIndex = 0;
```
Does it mean InSequence starts with 1 (second dot) instead of first dot?

yombunker replied:
yeah that's a bug i realized yesterday, i'll fix it in a moment

uDevel reviewed (commented):
Suggestions

yombunker committed:
- Fixing PR comments
- Adding unit test for the sequence generators
  - Except for the Random ones, cause those would be flaky tests

yombunker commented:
Ok, i already fixed the comments from the PR, and added a few Unit Tests to make sure the sequence are working as expected, for the Random one i couldn't imagine how a test would look like for them :P

yombunker committed:
- Chaning the name in the template

uDevel commented on typing-indicator/src/main/java/com/udevel/widgetlab/sequences/CircularSequence.java:
```diff
@@ -5,15 +5,43 @@
+
+        if (currentIndex >= numberOfElements) {
+            clockWise = false;
+            currentIndex = numberOfElements > 1 ? numberOfElements - 2 : 0;
```
this can be numberOfElements  > 2, it would be the same.

uDevel commented on typing-indicator/src/main/java/com/udevel/widgetlab/sequences/ReverseSequence.java:
```diff
@@ -5,9 +5,12 @@
  */
 
 public class ReverseSequence implements SequenceGenerator {
+    private int currentIndex = -1;
```
Can this be in abstract class?

yombunker replied:
What would be the benefit?

uDevel reviewed (approved):
Just 1 suggestion
